### ISSUE TITLE
clone

### ISSUE THREAD

**Issue by brankokaucic** [mpdf/mpdf#243]
Hi,

I have a problem with cloning of MPDF object. Cloning is used for checking if adding additional content with writeHTML will set Y on cloned object over some specific value. It works fine with 6.0, problems are with 6.1.0 (regardles if I use PHP 7 or PHP 5.6.19).

Used code is
$T = '.......'; // first part of content (header of table plus some rows)
while (....) {
  $T3 = '......'; // i-th row in table

  $mpdf2 = clone $mpdf;
  writeLog("mpdf->y=" . $mpdf->y . "; mpdf2->y=" . $mpdf2->y);

  $T2 = $T . $T3 . '< /table>';
  $mpdf2->WriteHTML($T2);
  writeLog("mpdf->y=" . $mpdf->y . "; mpdf2->y=" . $mpdf2->y);
}
///
In 6.0 I get these lines in log:
mpdf->y=; mpdf2->y=
mpdf->y=; mpdf2->y=122.63166666667

mpdf->y=; mpdf2->y=
mpdf->y=; mpdf2->y=126.79

mpdf->y=; mpdf2->y=
mpdf->y=; mpdf2->y=130.94833333333

mpdf->y=; mpdf2->y=
mpdf->y=; mpdf2->y=141.10666666667

mpdf->y=; mpdf2->y=
mpdf->y=; mpdf2->y=148.265

In 6.1.0 I get these lines in log:
mpdf->y=; mpdf2->y=
mpdf->y=20.191666666667; mpdf2->y=76.926584

mpdf->y=20.191666666667; mpdf2->y=20.191666666667
mpdf->y=41.541666666667; mpdf2->y=81.305552

mpdf->y=41.541666666667; mpdf2->y=41.541666666667
mpdf->y=64.05; mpdf2->y=85.68452

mpdf->y=64.05; mpdf2->y=64.05
mpdf->y=87.716666666667; mpdf2->y=85.68452

mpdf->y=87.716666666667; mpdf2->y=87.716666666667
mpdf->y=112.54166666667; mpdf2->y=90.063488

Is cloning functionality in 6.1.0 different than in 6.0? 

Regards,
Branko

**Comment by finwe** [comment]
Certainly not intentionally. I will look into it when I have some time. Can you please post more complete example code?

**Comment by brankokaucic** [comment]
Hi,
thanks for the reply. I have prepared example below.
Many many thanks in advance!
Branko

<?php

function writeLog($iT) {

    $Filename = "log.txt";
    $F = @fopen($Filename, "a");
    if ($F === null || $F == "")
        return;
    @fwrite($F, "$iT\n");
    @fclose($F);
}

$Format = 'A4-L';
$Filename = 'test.pdf';

$DocHeader = '<table width="100%" style="border-collapse: collapse;"><tr><td class="border" style="width: 20%; text-align: center;">LOGO</td><td class="border" style="width: 80%; text-align: center;">TITLE</td></tr></table>';

$DocFooter = '<table style="width: 100%; border-top: 0.1mm solid #000000; border-collapse: collapse;"><tr><td style="width: 40%; text-align: left;">LEFT</td><td style="width: 20%; text-align: center;">CENTER</td><td style="width: 40%; text-align: left;">RIGHT</td></tr></table>';

require("../../mpdf/mpdf.php");
$mpdf = new mPDF('utf-8', $Format, '', '', 10, 10, 20, 10, 5, 5);

$mpdf->SetHTMLHeader($DocHeader);
$mpdf->SetHTMLFooter($DocFooter);

$mpdf->SetDisplayMode('fullpage');

$THeader = '';
$THeader .= '<table style="width: 100%;" border="1">';
$THeader .= '<tr>';
$THeader .= '<th>ROWNUM</th>';
$THeader .= '<th>CELL 1</th>';
$THeader .= '<th>CELL 2</th>';
$THeader .= '<th>CELL 3</th>';
$THeader .= '</tr>';

$T = $THeader;
for ($i = 0; $i < 10; $i++) {

    $TRow = '';
    $TRow .= '<tr>';
    $TRow .= '<td>#</td>';
    $TRow .= '<td>VALUE 1</td>';
    $TRow .= '<td>VALUE 2</td>';
    $TRow .= '<td>VALUE 3</td>';
    $TRow .= '</tr>';

    $mpdf2 = clone $mpdf;
    writeLog("(1) i=$i; mpdf->y=" . $mpdf->y . "; mpdf2->y=" . $mpdf2->y);

    $T2 = $T . $TRow . '</table>';
    $mpdf2->WriteHTML($T2);
    writeLog("(1) i=$i; mpdf->y=" . $mpdf->y . "; mpdf2->y=" . $mpdf2->y);

    if ($mpdf2->y > 50) {
        writeLog("(3) over y limit");

        $mpdf->WriteHTML($T . '</table>');
        $mpdf->AddPage();
        $T = $THeader;
        $i--;
    } else {
        $T .= $TRow;
    }
}

$mpdf->WriteHTML($T . '</table>');

$mpdf->Output($Filename, 'I');
exit(0);
?>

**Comment by brankokaucic** [comment]
Can't upload ZIP file to comment, I hope that the example from previous post is readable enough.

B

**Comment by brankokaucic** [comment]
Hi,
was example useful? Should I provide more details about that issue?


Best regards,
Branko

**Comment by brankokaucic** [comment]
Hi,

any update on this issue? Is it maybe related to some constructor issue that was posted a while ago?

Best regards,
Branko

**Comment by finwe** [comment]
No, sorry, I have no time now to deal with that and I cannot give you an ETA either. There are much more important issues (for which I have no time either).

**Comment by kudm761** [comment]
Hi, guys, I investigated the problem.
`When an object is cloned, PHP will perform a shallow copy of all of the object's properties. Any properties that are references to other variables will remain references.`

When you clone mpdf object, its object properties are not cloned and they still the same.
So when you  do something with cloned object you impact on properties of original object, because they are actually references. In this way you get unexpected behaviour.

I made a pull-request #414 to allow cloning, but I am not sure that it is a good idea.
